src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1127,221 +1114,289 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMCluster> const& m)
+        auto reply{shardStore->getShardInfo()->makeMessage(app_)};
+        if (peerChainSz > 0)
+            *(reply.mutable_peerchain()) = m->peerchain();
+        send(std::make_shared<Message>(reply, protocol::mtPEER_SHARD_INFO_V2));
+    }
 
-        m->set_hops(m->hops() - 1);
-        if (m->hops() == 0)
-            m->set_lastlink(true);
+    if (m->relays() == 0)
+        return;
 
-        m->add_peerchain()->set_nodepubkey(
-            publicKey_.data(), publicKey_.size());
+    // Add peer to the peer chain
+    m->add_peerchain()->set_publickey(publicKey_.data(), publicKey_.size());
 
-        overlay_.foreach(send_if_not(
-            std::make_shared<Message>(*m, protocol::mtGET_PEER_SHARD_INFO),
-            match_peer(this)));
-    }
+    // Relay the request to peers, exclude the peer chain
+    m->set_relays(m->relays() - 1);
+    overlay_.foreach(send_if_not(
+        std::make_shared<Message>(*m, protocol::mtGET_PEER_SHARD_INFO_V2),
+        [&](std::shared_ptr<Peer> const& peer) {
+            return pubKeyChain.find(peer->getNodePublic()) != pubKeyChain.end();
+        }));
+
+    fee_ = Resource::feeMediumBurdenPeer;
 }
 
 void
-PeerImp::onMessage(std::shared_ptr<protocol::TMPeerShardInfo> const& m)
-{
+PeerImp::onMessage(std::shared_ptr<protocol::TMPeerShardInfoV2> const& m)
+{
+    // Find the earliest and latest shard indexes
+    auto const& db{app_.getNodeStore()};
+    auto const earliestShardIndex{db.earliestShardIndex()};
+    auto const latestShardIndex{[&]() -> std::optional<std::uint32_t> {
+        auto const curLedgerSeq{app_.getLedgerMaster().getCurrentLedgerIndex()};
+        if (curLedgerSeq >= db.earliestLedgerSeq())
+            return db.seqToShardIndex(curLedgerSeq);
+        return std::nullopt;
+    }()};
+
     auto badData = [&](std::string msg) {
         fee_ = Resource::feeBadData;
         JLOG(p_journal_.warn()) << msg;
     };
 
-    if (m->shardindexes().empty())
-        return badData("Missing shard indexes");
-    if (m->peerchain_size() > csHopLimit)
-        return badData("Invalid peer chain");
-    if (m->has_nodepubkey() && !publicKeyType(makeSlice(m->nodepubkey())))
-        return badData("Invalid public key");
+    // Used to create a digest and verify the message signature
+    Serializer s;
+    s.add32(HashPrefix::shardInfo);
 
-    // Check if the message should be forwarded to another peer
-    if (m->peerchain_size() > 0)
+    // Verify message creation time
+    NodeStore::ShardInfo shardInfo;
     {
-        // Get the Public key of the last link in the peer chain
-        auto const s{
-            makeSlice(m->peerchain(m->peerchain_size() - 1).nodepubkey())};
-        if (!publicKeyType(s))
-            return badData("Invalid pubKey");
-        PublicKey peerPubKey(s);
+        auto const timestamp{
+            NetClock::time_point{std::chrono::seconds{m->timestamp()}}};
+        auto const now{app_.timeKeeper().now()};
+        if (timestamp > now)
```

Comment:
`if (timestamp > (now + 5s))` to allow some peer clocks that are not synchronized?

Reply:
@pwang200 Good idea, thank you.